es_out: support Upstream Servers with configuration overriding

mabrarov opened the pull request:
Implementation of [Upstream](https://docs.fluentbit.io/manual/configuration/upstream_servers) feature for the Elasticsearch output plugin.

This pull request is based on pull request #1560 and Forward output plugin.

It was tested in a local setup with:

1. Fluent Bit without Upstream feature connected to a single node of Elasticsearch cluster consisting of 3 master-eligible/data and 1 coordinating nodes.

    Refer to [elastic-cluster](https://github.com/mabrarov/elastic-stack/tree/main/elastic-cluster) directory of [mabrarov/elastic-stack](https://github.com/mabrarov/elastic-stack) repository for Docker Compose project used to create target Elasticsearch cluster and Kibana.

    fluent-bit.conf Fluent Bit configuration file used for the test - refer to [fluent-bit-es/fluent-bit.conf](https://github.com/mabrarov/elastic-stack/blob/main/fluent-bit-es/fluent-bit.conf) and (same in YAML format) [fluent-bit-es/fluent-bit.yaml](https://github.com/mabrarov/elastic-stack/blob/main/fluent-bit-es/fluent-bit.yaml) in [mabrarov/elastic-stack](https://github.com/mabrarov/elastic-stack) repository.

    Debug log is available at [flb_es.log](https://drive.google.com/file/d/1NHLs1KLRu49yMjrg86uulvQeyhaCoCUd/view?usp=sharing).

1. Fluent Bit with Upstream feature connected to all Elasticsearch data nodes of Elasticsearch cluster consisting of 3 master-eligible/data and 1 coordinating nodes.

    Refer to [elastic-cluster](https://github.com/mabrarov/elastic-stack/tree/main/elastic-cluster) directory of [mabrarov/elastic-stack](https://github.com/mabrarov/elastic-stack) repository for Docker Compose project used to create target Elasticsearch cluster and Kibana.

    fluent-bit.conf Fluent Bit configuration file used for the test - refer to [fluent-bit-es-cluster/fluent-bit.conf](https://github.com/mabrarov/elastic-stack/blob/main/fluent-bit-es-cluster/fluent-bit.conf) and (same in YAML format) [fluent-bit-es-cluster/fluent-bit.yaml](https://github.com/mabrarov/elastic-stack/blob/main/fluent-bit-es-cluster/fluent-bit.yaml) in [mabrarov/elastic-stack](https://github.com/mabrarov/elastic-stack) repository.

    Debug log is available at [flb_es_upstream.log](https://drive.google.com/file/d/1lcdkvTza62scmW7eCBI86_J1r7e7XDQP/view?usp=sharing).

----

**Testing**

- [x] Example configuration files for the change can be found in [mabrarov/elastic-stack](https://github.com/mabrarov/elastic-stack) repository under [fluent-bit-es-cluster](https://github.com/mabrarov/elastic-stack/tree/main/fluent-bit-es-cluster) directory.
- [x] Debug log output from testing the change - see above.
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found - refer to [flb_run_code_analysis.log](https://drive.google.com/file/d/17R4zlWf140y8g1EUpIqr8Ax0CG0vWWV9/view?usp=sharing) for the output of command
    ```bash
    TEST_PRESET=valgrind SKIP_TESTS='flb-rt-out_forward flb-rt-out_td' ./run_code_analysis.sh
    ```
- [N/A] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [N/A] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**

- [x] Documentation required for this feature - refer to https://github.com/fluent/fluent-bit-docs/pull/1143.

**Backporting**

- [N/A] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * HA-aware Elasticsearch output with per-upstream/per-node configuration and upstream-aware formatter flush-context callbacks.

* **Integrations**
  * Cloud ID parsing for Elasticsearch endpoints.
  * Optional AWS SigV4 support with STS and unsigned-headers handling.

* **Configuration**
  * Consolidated ES property names, added upstream option, and corrected default Elasticsearch port to 9200.

* **Quality**
  * Ownership-aware resource handling and improved lifecycle/cleanup.

* **Tests**
  * Expanded upstream-focused tests covering formatter, flush-context, and config scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mabrarov committed:
out_es: overriding plugin parameters within upstream node

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov commented on plugins/out_es/es.h:
```diff
@@ -31,24 +34,24 @@
-#define FLB_ES_WRITE_OP_UPSERT    "upsert"
 
-struct flb_elasticsearch {
+struct flb_elasticsearch_config {
```
The whole idea / implementation and naming of new types follows Upstream support in Forward output plugin.

mabrarov commented on plugins/out_es/es_conf_parse.c:
```diff
@@ -0,0 +1,319 @@
+        return 0;
+    }
+
+    toks = flb_utils_split((const char *)cloud_auth, ':', -1);
```
Not:

```suggestion
    toks = flb_utils_split((const char *) cloud_auth, ':', -1);
```

because https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md requires to follow https://httpd.apache.org/dev/styleguide.html which requires not put space in case of type cast:

> **The Guidelines**
> ...
> There is no whitespace between a cast and the item modified (e.g., `"(int)j"` and not `"(int) j"`).

mabrarov commented:
Hi reviewers,

Is it possible to approve only workflow for this pull request, so that automated checks and build can start?

Thank you.

PettitWesley commented:
@mabrarov sure

mabrarov commented:
Hi @PettitWesley,

It looks like all failed checks are around run-macos-unit-tests jobs and caused by the following failed unit tests:

1. flb-rt-in_event_test
1. flb-rt-out_tcp

I feel like other pull requests have the same issues, i.e. it doesn't seem that the failed checks are caused by this pull request changes.

Thank you.

mabrarov commented on plugins/out_es/es_conf.c:
```diff
@@ -17,502 +17,818 @@
+    }
+    tmp = flb_upstream_node_get_property(FLB_ES_CONFIG_PROPERTY_LOGSTASH_PREFIX, node);
+    if (tmp) {
+        ec->logstash_prefix = (char *)tmp;
```
Type cast is illegal. Should use `flb_sds_create` instead. Working on fix.

mabrarov replied:
Fixed.

mabrarov commented:
Hi @PettitWesley,

Is it possible to trigger automated workflow (build) for this pull request one more time? I found & fixed one issue and added tests for the new code since last build happened.

Thank you.

mabrarov commented:
Hi dear reviewers,

Is it possible to get this pull request reviewed / accepted sooner? Is there something pending / waiting from my side to start review?

Thank you.

mabrarov commented:
Hi @PettitWesley  and @edsiper,

It feels like you are code owners for Elasticsearch output plugin. Is there something pending / waiting from my side to start review of this pull request? This new feature was requested 4 years ago and I feel it is something which multiple users of Fluent Bit (not just my team) would like to have.

Thank you.

mabrarov commented:
Hi @PettitWesley and @edsiper,

Is there a chance to get this pull request reviewed (and eventually accepted)? Please let me know if something is missing and blocking that.

Thank you.

116davinder commented:
Please someone review this, I also want this feature to merged and released.
thank you!

dioniseo commented:
+1

Hello reviewers, could you please review this PR?

This feature also will be very useful in our project and very useful feature to support HA.

Thank you!

mabrarov commented:
Just FYI, my team successfully tested replacement of following log shipping solution:

Fluent Bit DaemonSet in K8s (Forward output plugin with Upstream Servers, Upstream node per Fluentd instance) → Fluentd (2 instances with fluent-plugin-elasticsearch configured to connect to all nodes of Elasticsearch cluster) → Elasticsearch cluster (3+ nodes)

with:

Fluent Bit DaemonSet in K8s (Docker image built from source branch of this pull request, Elasticsearch output plugin with Upstream Servers, Upstream node per Elasticsearch node) → Elasticsearch cluster (3+ nodes)

For my team it means (this pull request means) a possibility to get rid of 7-9 Fluentd instances (where each Fluentd instance occupies a single VM).

mabrarov commented:
This pull request is still actual and up-to-date (ready for merging, tested with changes coming from target branch).

AzureLeMoon commented:
Hello, @PettitWesley and @edsiper.

i hope you guys can take a look at this pr and merge it sooner, it's a highly requested feature with its mentioned in the official docs as well.

mabrarov commented:
Hi

It looks like there is one more enhancement which this pull request brings comparing to approach "Fluent Bit (Forward output plugin with Upstream Servers) → multiple Fluentd instances → multiple Elasticsearch instances of a single Elasticsearch cluster":

1. Fluentd can generate unique ID of Elasticsearch document before inserting data into Elasticsearch. This approach can help to avoid duplicates when re-sending data to Elasticsearch at Fluentd side. Refer to https://github.com/uken/fluent-plugin-elasticsearch?tab=readme-ov-file#generate-hash-id.
1. Unfortunately, `elasticsearch_genid` Fluentd filter plugin won't help to avoid duplicate records when Fluent Bit re-sends data to Fluentd.
1. With this pull request we can generate unique ID of Elasticsearch document at Fluent Bit side - refer to `Generate_ID` key of [Elasticsearch Fluent Bit output plugin configuration](https://docs.fluentbit.io/manual/v/dev-2.2/pipeline/outputs/elasticsearch). This approach (Fluent Bit built from this pull request; Fluent Bit (Elasticsearch output plugin with Upstream Servers and `Generate_ID` configuration key set to `On`) → multiple Elasticsearch instances of a single Elasticsearch cluster) fully eliminates possibility to get duplicated records at Elasticsearch side due to re-sending of data.

mabrarov commented:
Hi @PettitWesley,

Could you (or someone from maintainers / reviewers) please approve workflow for this pull request, so that the automated checks and build of this pull request can start?

Thank you.

mabrarov commented:
FYI, the competitor of Fluent Bit - Vector - supports multiple Elasticsearch hosts in its [elastcisearch](https://vector.dev/docs/reference/configuration/sinks/elasticsearch/) sink. Refer to https://vector.dev/docs/reference/configuration/sinks/elasticsearch/#endpoints. Note that it looks like Vector uses smart approach (not like this pull request) when deciding what Elasticsearch endpoint to send data, because sink configuration supports [healthcheck](https://vector.dev/docs/reference/configuration/sinks/elasticsearch/#healthcheck). Just like Fluentd [Elasticsearch plugin](https://github.com/uken/fluent-plugin-elasticsearch).

cosmo0920 commented on plugins/out_es/es_conf_parse.c:
```diff
@@ -0,0 +1,338 @@
+        port = colon+1;
+    }
+
+    strcpy(cloud_host_buf, host);
+    strcat(cloud_host_buf, ".");
+    strcat(cloud_host_buf, region);
+    if (port != NULL) {
+        strcat(cloud_host_buf, ":");
+        strcat(cloud_host_buf, port);
+    }
```
We need to use strict length version of strncat and strcpy at least.
Or, Fluent Bit offers the flb_sds function APIs which can handling more safely to process strings.
Could you use them if you could do?

mabrarov replied:
This part is original code - refer to https://github.com/fluent/fluent-bit/blob/master/plugins/out_es/es_conf.c#L92-L95 - moved to a new location. More changes we introduce into this pull request - harder it will be accept / test this pull request. I'm not against of fixing issues of original code in this pull request too (in dedicated commit(s)), but I'd like to ensure it is really needed and requesters of fixes realize the source of these issues.

cosmo0920 reviewed (changes requested):
I also found the dangerous functions usages from libc.
The added patch is basically reasonable at the point of codes.
I'll take more time to review this PR tomorrow.

mabrarov commented on plugins/out_es/es_conf_parse.c:
```diff
@@ -0,0 +1,338 @@
+        flb_plg_error(ctx->ins, "cannot decode cloud_id");
+        return NULL;
+    }
+    region = strtok(buf, dollar);
+    if (region == NULL) {
+        return NULL;
+    }
+    host = strtok(NULL, dollar);
+    if (host == NULL) {
+        return NULL;
+    }
```
It's original code - refer to https://github.com/fluent/fluent-bit/blob/master/plugins/out_es/es_conf.c#L67-L74 - moved to a new location. We can try to fix it, but that way we will spend time in this pull request fixing all issues of original code (which is not goal of this pull request). Sorry, but I am not sure this pull request is the best place / time for doing that.

cosmo0920 reviewed (commented):
I confirmed that this PR is working well for HA settings.
Basically, `:+1:` for this.
But there is a few of dangerous function usages to be existing.

cosmo0920 commented on src/flb_lib.c:
```diff
@@ -549,6 +549,18 @@ int flb_output_set_test(flb_ctx_t *ctx, int ffd, char *test_name,
                         void (*out_callback) (void *, int, int, void *, size_t, void *),
                         void *out_callback_data,
                         void *test_ctx)
+{
+    return flb_output_set_test_with_ctx_callback(ctx, ffd, test_name, out_callback,
+                                                 out_callback_data, test_ctx, NULL);
+}
+
+int flb_output_set_test_with_ctx_callback(flb_ctx_t *ctx, int ffd, char *test_name,
+                        void (*out_callback) (void *, int, int, void *, size_t, void *),
+                        void *out_callback_data,
+                        void *test_ctx,
+                        void *(*test_ctx_callback) (struct flb_config *,
+                                                    struct flb_input_instance *,
+                                                    void *, void *))
```
This change shouldn't be good for maintanancibility.

Instead, how about using the following style?

```c
int flb_output_set_test(flb_ctx_t *ctx, int ffd, char *test_name,
                        void (*out_callback) (void *, int, int, void *, size_t, void *),
                        void *out_callback_data,
                        void *test_ctx)
{
    struct flb_output_instance *o_ins;

    o_ins = out_instance_get(ctx, ffd);
    if (!o_ins) {
        return -1;
    }

    /*
     * Enabling a test, set the output instance in 'test' mode, so no real
     * flush callback is invoked, only the desired implemented test.
     */

    /* Formatter test */
    if (strcmp(test_name, "formatter") == 0) {
        o_ins->test_mode = FLB_TRUE;
        o_ins->test_formatter.rt_ctx = ctx;
        o_ins->test_formatter.rt_ffd = ffd;
        o_ins->test_formatter.rt_out_callback = out_callback;
        o_ins->test_formatter.rt_data = out_callback_data;
        o_ins->test_formatter.flush_ctx = test_ctx;
        o_ins->test_formatter.flush_ctx_callback = NULL;
    }
    else {
        return -1;
    }

    return 0;
}

int flb_output_set_test_ctx_callback(flb_ctx_t *ctx, int ffd, char *test_name,
                                     void *(*test_ctx_callback) (struct flb_config *,
                                                                 struct flb_input_instance *,
                                                                 void *, void *))
{
    struct flb_output_instance *o_ins;

    o_ins = out_instance_get(ctx, ffd);
    if (!o_ins) {
        return -1;
    }

    /*
     * Enabling a test, set the output instance in 'test' mode, so no real
     * flush callback is invoked, only the desired implemented test.
     */

    /* Formatter test */
    if (strcmp(test_name, "formatter") == 0) {
        o_ins->test_formatter.rt_ctx = ctx;
        o_ins->test_formatter.rt_ffd = ffd;
        o_ins->test_formatter.flush_ctx_callback = test_ctx_callback;
    }
    else {
        return -1;
    }

    return 0;
}
```

Plus, the corresponsed C header difinitions like:

```c
FLB_EXPORT int flb_output_set_test(flb_ctx_t *ctx, int ffd, char *test_name,
                                   void (*out_callback) (void *, int, int,
                                                         void *, size_t, void *),
                                   void *out_callback_data,
                                   void *test_ctx);
FLB_EXPORT int flb_output_set_ctx_callback(flb_ctx_t *ctx, int ffd, char *test_name,
                                           void *(*test_ctx_callback) (struct flb_config *,
                                                                       struct flb_input_instance *,
                                                                       void *, void *));
```

Adding simply the new argument is not good idea to maintain code base.

If you use these suggested functions, you also need to change for unit testing file of out_elasticsearch plugin.

mabrarov replied:
I'm sorry, but the tests for Elasticsearch output plugin need both callbacks at the same time - not just one of `out_callback` or `test_ctx_callback`. The `out_callback` is used in the tests for validation / assertion, while `test_ctx_callback` is used to extract the right value used later as `out_context` parameter of Elasticsearch output plugin's `cb_es_flush` function (which is `cb_flush` callback of Fluent Bit output plugin). This change is not just about adding one more parameter, but is about providing tests with flexibility to make tested plugin acting just like it does during its work, where the "optional flush context" parameter of formatter function is not a constant but is controlled by `cb_flush` callback.

The same approach could be used for the tests for Forward output plugin, i.e. this change is not just for Elasticsearch plugin. The only reason tests for Forward output plugin doesn't use this approach is because of https://github.com/fluent/fluent-bit/blob/master/plugins/out_forward/forward_format.c#L563 where, IMHO, `flb_forward_format` function breaks its responsibility area just to not break the tests :)

mabrarov replied:
... maybe if we change `flb_output_set_test` to something like (removing line `o_ins->test_formatter.flush_ctx_callback = NULL;`):

```cpp
int flb_output_set_test(flb_ctx_t *ctx, int ffd, char *test_name,
                        void (*out_callback) (void *, int, int, void *, size_t, void *),
                        void *out_callback_data,
                        void *test_ctx)
{
    struct flb_output_instance *o_ins;

    o_ins = out_instance_get(ctx, ffd);
    if (!o_ins) {
        return -1;
    }

    /*
     * Enabling a test, set the output instance in 'test' mode, so no real
     * flush callback is invoked, only the desired implemented test.
     */

    /* Formatter test */
    if (strcmp(test_name, "formatter") == 0) {
        o_ins->test_mode = FLB_TRUE;
        o_ins->test_formatter.rt_ctx = ctx;
        o_ins->test_formatter.rt_ffd = ffd;
        o_ins->test_formatter.rt_out_callback = out_callback;
        o_ins->test_formatter.rt_data = out_callback_data;
        o_ins->test_formatter.flush_ctx = test_ctx;
    }
    else {
        return -1;
    }

    return 0;
}
```

Then we can use `flb_output_set_test` and `flb_output_set_ctx_callback` together (one-by-one) achieving the same as we have right now in this pull request. But due to `test_formatter` is not initialized somewhere else, this approach will require one more function to fill `test_formatter.flush_ctx` and `test_formatter.flush_ctx_callback` with `NULL` which sounds like a complication (will require a lot of changes in existing tests - we will need to call that init function everywhere before `flb_output_set_test` function).

mabrarov replied:
Implemented in 7ee017f7080c34ec15e11e06e6cdc556ba55f1f4. I called new function `flb_output_set_test_flush_ctx_callback` instead of `flb_output_set_ctx_callback`. Thank you for a good suggestion!

cosmo0920 commented on plugins/out_es/es.h:
```diff
@@ -66,14 +70,17 @@ struct flb_elasticsearch {
     char *aws_sts_endpoint;
     char *aws_profile;
     struct flb_aws_provider *aws_provider;
+    int own_aws_provider;
```
We shouldn't use this pattern. This is because when initializing struct typed values or basic type of values, we can distinguish whether allocated or not.

mabrarov replied:
Answered in [another comment](https://github.com/fluent/fluent-bit/pull/7608#issuecomment-2526291616). `own_xxx` fields are used to deallocate data shared b/w plugin configuration and upstream node configuration just once.

cosmo0920 commented on plugins/out_es/es_conf.c:
```diff
@@ -17,521 +17,899 @@
+
+    /* Copy base configuration */
+    *ec = *base;
+    ec->own_index = FLB_FALSE;
+    ec->own_type = FLB_FALSE;
+    ec->own_cloud_user = FLB_FALSE;
+    ec->own_cloud_passwd = FLB_FALSE;
```
We have to initialize with NULL instead of filling these variables with FLB_FALSE.

mabrarov replied:
Answered in [another comment](https://github.com/fluent/fluent-bit/pull/7608#issuecomment-2526291616). `own_xxx` fields are used to deallocate data shared b/w plugin configuration and upstream node configuration just once.

cosmo0920 replied:
> Answered in [another comment](https://github.com/fluent/fluent-bit/pull/7608#issuecomment-2526291616). `own_xxx` fields are used to deallocate data shared b/w plugin configuration and upstream node configuration just once.

We use C not C++. So, we can use initializing with NULL and free pattern. If you use C++, your comments is correct.
However, C does not have ownership on allocated memory. Thanks.

cosmo0920 commented on plugins/out_es/es_conf.c:
```diff
@@ -17,521 +17,899 @@
+        flb_sds_destroy(ec->tag_key);
+    }
+
+    if (ec->ra_id_key && ec->own_ra_id_key == FLB_TRUE) {
```
We can distinguish whether ec->ra_id_key with if condition when initialized with NULL on initialize.

mabrarov replied:
Answered in [another comment](https://github.com/fluent/fluent-bit/pull/7608#issuecomment-2526291616). `own_xxx` fields are used to deallocate data shared b/w plugin configuration and upstream node configuration just once.

cosmo0920 reviewed (changes requested):
Checking whether initialized or not with `own_xxx` is bad idea and it's too defensive patterns for our code style.
Initializing planning to be allocated variable should be initialized with NULL and checking whether it was allocated or not with `if (ec->xxx) { ... }` pattern.
NULL checking pattern should be enough to release allocated resources. This pattern could not be working properly when lacking of initializing with NULL on initialize.

mabrarov commented:
Hi @cosmo0920,

> Checking whether initialized or not with `own_xxx` is bad idea

`own_xxx` is not about initialization, but is about ownership. Some data (to simplify the code and to reduce memory footprint) is shared b/w output plugin configuration and upstream node configuration, so it is required to know what entity owns the data to deallocate the data just once.

>Initializing planning to be allocated variable should be initialized with NULL and checking whether it was allocated or not with `if (ec->xxx) { ... }` pattern. NULL checking pattern should be enough to release allocated resources.

`NULL` check is not required for `flb_free` function, because it uses C standard library [free](https://en.cppreference.com/w/c/memory/free) which accepts `NULL`:

Thank you.

mabrarov commented:
Hi @cosmo0920,

Regarding your [comment](https://github.com/fluent/fluent-bit/pull/7608#discussion_r1875169852) - the ownership concept have no relationship with language. If we remove all `own_xxx` fields and run Fluent Bit with upstream servers (refer to [fluent-bit-es-cluster/fluent-bit.conf](https://github.com/mabrarov/elastic-stack/blob/main/fluent-bit-es-cluster/fluent-bit.conf) in description of this pull request) and Valgrind, then we will get double deallocation error. Please check the code one more time and thank you for the time you spend on review of these changes.

cosmo0920 commented:
> Hi @cosmo0920,
> 
> Regarding your [comment](https://github.com/fluent/fluent-bit/pull/7608#discussion_r1875169852) - the ownership concept have no relationship with language. If we remove all `own_xxx` fields and run Fluent Bit with upstream servers (refer to [fluent-bit-es-cluster/fluent-bit.conf](https://github.com/mabrarov/elastic-stack/blob/main/fluent-bit-es-cluster/fluent-bit.conf) in description of this pull request) and Valgrind, then we will get double deallocation error, because the same value (same address) is stored in multiple pointers. Please check the code one more time and thank you for the time you spend on review of these changes.

In our code base, there is no own_xxx fields even if flb_upstream using plugins such as out_forward.
This indicates that your code style does not fit for ours. Thanks.

cosmo0920 commented:
> Hi @cosmo0920,
> 
> Regarding your [comment](https://github.com/fluent/fluent-bit/pull/7608#issuecomment-2527720627).
> 
> The only Fluent Bit output plugin supporting upstream servers (before this pull request) is Forward output plugin and that plugin seems to not support overriding common configuration of plugin per each upstream node. That's the reason Forward output plugin doesn't need to care about ownership of configuration data - each upstream node just use its own set of configuration data. User of Forward output plugin has to specify (duplicate) plugin options at upstream node level otherwise default value will be used (e.g. `shared_key`, `username` and `password` options), i.e there is no place to specify configuration which is common for all upstream nodes of specific Forward output plugin instance. This approach is not going to work well with Elasticsearch output plugin, because it has too many options to duplicate them per each upstream node.
> 
> Refer to the linked [fluent/fluent-bit-docs#1143](https://github.com/fluent/fluent-bit-docs/pull/1143) for the changes in documentation for Elasticsearch output plugin which clearly state what options can be overridden at upstream node level (or configuration of output plugin is used otherwise).
> 
> Thank you.

Hi,
If your suggestion is true, we need to implement overriding mechanism on configuration on upstream nodes.
With the current implementation, it is too deviated from our coding rules.
From the own_xxx pattern, there is not good smells.

mabrarov commented:
After recent rebase tests failed. It looks like it is caused by upgraded version of CMake - refer to https://github.com/actions/runner-images/issues/11926

mabrarov commented:
> After recent rebase tests failed. It looks like it is caused by upgraded version of CMake - refer to [actions/runner-images#11926](https://github.com/actions/runner-images/issues/11926)

Fixed CMake version in CI configuration in https://github.com/fluent/fluent-bit/pull/10178 which was cherry-picked into this PR in 69861bd076026fae35ef20a68572b9e3fa178c4b.

mabrarov commented on .github/workflows/unit-tests.yaml:
```diff
@@ -35,22 +35,25 @@ on:
           - "-DFLB_JEMALLOC=On"
           - "-DFLB_JEMALLOC=Off"
-          - "-DFLB_SMALL=On"
+          # Fails even on master branch after migrated from GCC 7 to GCC 9 in https://github.com/fluent/fluent-bit/pull/10230
```
Just to test https://github.com/fluent/fluent-bit/pull/10230

mabrarov commented on .github/workflows/unit-tests.yaml:
```diff
@@ -35,22 +35,25 @@ on:
           - "-DFLB_JEMALLOC=On"
           - "-DFLB_JEMALLOC=Off"
-          - "-DFLB_SMALL=On"
+          # Fails even on master branch after migrated from GCC 7 to GCC 9 in https://github.com/fluent/fluent-bit/pull/10230
+          #- "-DFLB_SMALL=On"
```
Just to test https://github.com/fluent/fluent-bit/pull/10230

mabrarov commented on .github/workflows/unit-tests.yaml:
```diff
@@ -63,14 +65,24 @@ jobs:
             compiler:
               cc: clang
               cxx: clang++
+          # This build variant fails even on master branch after migration from GCC 7 to GCC 9 in https://github.com/fluent/fluent-bit/pull/10230
+          - flb_option: "-DFLB_SMALL=On"
+            compiler:
+              cc: gcc
+              cxx: g++
```
Just to test https://github.com/fluent/fluent-bit/pull/10230

mabrarov replied:
Should be fixed in https://github.com/fluent/fluent-bit/pull/10234

patrick-stephens commented:
@mabrarov / @cosmo0920 are the issues covered here? There's a lot of stuff in the conversation so not sure if the changes have been actioned.

cosmo0920 reviewed (changes requested):
We might want to use slightly different approach for achieving this.

This is because own_* flags are frequently appeared in this PR.
So, we wanted to wrap them with the following struct like:

```c
/* A wrapper for flb_sds_t that tracks memory ownership */
typedef struct {
    flb_sds_t sds;   /* The actual string data */
    int       owned; /* A flag indicating if this struct owns the memory */
} flb_es_sds_t;
```

Then, the wrapped members could be:

```c
/* In es.h, struct flb_elasticsearch_config */
struct flb_elasticsearch_config {
    /* ... other properties ... */

    flb_es_sds_t index;
    flb_es_sds_t type;
    flb_es_sds_t http_user;
    flb_es_sds_t logstash_prefix;
    flb_es_sds_t time_key;

    /* ... etc ... */
};
```

For just initialization, it could be:

```c
static inline void es_sds_init(flb_es_sds_t *dest, flb_es_sds_t *src)
{
    dest->sds = src->sds;
    dest->owned = FLB_FALSE; /* The new struct does not own the shared memory */
}
```

For setting up a value, it could be:

```c
static inline int es_sds_set(flb_es_sds_t *wrapper, const char *value)
{
    /* If we previously owned a different string, free it first */
    if (wrapper->owned && wrapper->sds) {
        flb_sds_destroy(wrapper->sds);
    }

    wrapper->sds = flb_sds_create(value);
    if (!wrapper->sds) {
        return -1;
    }
    wrapper->owned = FLB_TRUE;
    return 0;
}
```

The teardown operation of these values could be:

```c
static inline void es_sds_destroy(flb_es_sds_t *wrapper)
{
    if (wrapper->owned && wrapper->sds) {
        flb_sds_destroy(wrapper->sds);
    }
    wrapper->sds = NULL;
    wrapper->owned = FLB_FALSE;
}
```

Then, just teardown operation could be:

```c
es_sds_destroy(&ec->index);
es_sds_destroy(&ec->type);
/* ... etc ... */
```

These changes could be cleaner way to handle shallow copy.
We would love to use this type of wrapper to encapsulate for these type of complex operations.

mabrarov commented:
Hi @cosmo0920,

Regarding:

> We might want to use slightly different approach for achieving this.
> 
> This is because own_* flags are frequently appeared in this PR. So, we wanted to wrap them with the following struct like:
> 
> ```c
> /* A wrapper for flb_sds_t that tracks memory ownership */
> typedef struct {
>     flb_sds_t sds;   /* The actual string data */
>     int       owned; /* A flag indicating if this struct owns the memory */
> } flb_es_sds_t;
> ```

I like this approach, but there might be not just `flb_sds_t` type which will be wrapped if we decide to use this approach (C does not have C++ templates 😄). Let me check and try during this week.

Thank you.

cosmo0920 commented:
> Regarding:
> 
> > We might want to use slightly different approach for achieving this.
> > This is because own_* flags are frequently appeared in this PR. So, we wanted to wrap them with the following struct like:
> > ```c
> > /* A wrapper for flb_sds_t that tracks memory ownership */
> > typedef struct {
> >     flb_sds_t sds;   /* The actual string data */
> >     int       owned; /* A flag indicating if this struct owns the memory */
> > } flb_es_sds_t;
> > ```
> 
> I like this approach, but there might be not just `flb_sds_t` type which will be wrapped (I'm looking at `elasticsearch_config_destroy` function in plugins/out_es/es_conf.c) if we decide to use this approach (C does not have C++ templates 😄).

Hi,
Thanks for the response. I mean, I'm not objection for the defining each of necessary functions for wrapping up for shallow copies. Just for wanted to conceal own_* flags.

mabrarov commented:
Hi @cosmo0920,

Here is one more thing I found.

Some of configuration options - like `flb_elasticsearch_config::index` defined in plugins/out_es/es.h - are loaded using `flb_config_map`, which instance is defined at plugins/out_es/es.c - refer to https://github.com/fluent/fluent-bit/blob/42e29d6807c450e6f4a263aa51e54be150d35796/plugins/out_es/es.c#L1077

If I use custom type for such configuration options (fields of `flb_elasticsearch_config` struct) - like `flb_es_char_ptr` struct for `index`:

```c
struct flb_es_char_ptr {
    char *ptr;
    int   owned;
};

struct flb_elasticsearch_config {
    /* Elasticsearch index (database) and type (table) */
    struct flb_es_char_ptr index;
    ...
};
```

then it will complicate usage of `flb_config_map` too. Like (I'm not sure this code is correct):

```c
static struct flb_config_map config_map[] = {
    {
     FLB_CONFIG_MAP_STR, FLB_ES_CONFIG_PROPERTY_INDEX, FLB_ES_DEFAULT_INDEX,
     0, FLB_TRUE, offsetof(struct flb_elasticsearch_config, index) + offsetof(struct flb_es_char_ptr, ptr),
     "Set an index name"
    },
   ...
```

Anyway, I need to try it first. Just some thoughts.

Thank you.

mabrarov commented on tests/runtime/out_elasticsearch.c:
```diff
@@ -1,11 +1,117 @@
+static const char * const es_upstream_section_property_prefix = "    ";
+static const char * const es_upstream_section_value_prefix    = " ";
+
+static const char *create_upstream_conf_file(const char *first_property, ...)
+{
+    char *upstream_conf_filename;
+    FILE *upstream_conf_file;
+    int ret;
+    const char *arg;
+    int arg_idx;
+    va_list args;
+
+    upstream_conf_filename = tmpnam(NULL);
+    if (upstream_conf_filename == NULL) {
+        return NULL;
+    }
+
+    upstream_conf_file = fopen(upstream_conf_filename, "w");
+    if (upstream_conf_file == NULL) {
+        return NULL;
+    }
+
```
`mkstemp` doesn't seem to be a part of C standard. Considering `tmpnam` is used in test only, it can be acceptable, but the use of internal static buffer (`NULL` as argument of `tmpnam`) doesn't look OK.

mabrarov commented on plugins/out_es/es.c:
```diff
@@ -871,25 +909,25 @@ static void cb_es_flush(struct flb_event_chunk *event_chunk,
     /* Compose HTTP Client request */
-    c = flb_http_client(u_conn, FLB_HTTP_POST, ctx->uri,
+    c = flb_http_client(u_conn, FLB_HTTP_POST, ec->uri,
                         pack, pack_size, NULL, 0, NULL, 0);
 
-    flb_http_buffer_size(c, ctx->buffer_size);
+    flb_http_buffer_size(c, ec->buffer_size);
 
```
Existing issue. It is too risky (to heavy for this PR) to fix it.

cosmo0920 reviewed (changes requested):
As this feature grows, having these definitions spread out could become challenging to manage. Perhaps we could unify them to make it more robust?

So, I thought that generalizing creations and destruction could be unified with C macro. Like:

```c
#define OUT_ES_OWNED_MEMBER(type, name) \
    type name;                          \
    int own_##name;
```

and 

```c
#define OUT_ES_DESTROY_OWNED_MEMBER(config_ptr, member_name, destroy_func) \
    do {                                                            \
        if ((config_ptr)->own_##member_name && (config_ptr)->member_name) { \
            destroy_func((config_ptr)->member_name);                \
        }                                                           \
    } while (0)
```

These macros encapsulate the details of implementations that we frequently pointed out for fragility and maintainability.

If we use this type of macros, we could use like:

```c
struct flb_elasticsearch_config {
    /* Elasticsearch index (database) and type (table) */
    OUT_ES_DECLARE_OWNED_MEMBER(char *, index);
    OUT_ES_DECLARE_OWNED_MEMBER(char *, type);
    int suppress_type_name;

    /* Elastic Cloud Auth */
    OUT_ES_DECLARE_OWNED_MEMBER(char *, cloud_user);
    OUT_ES_DECLARE_OWNED_MEMBER(char *, cloud_passwd);

#ifdef FLB_HAVE_AWS
    /* ... other AWS members ... */
    OUT_ES_DECLARE_OWNED_MEMBER(struct flb_aws_provider *, aws_provider);
    OUT_ES_DECLARE_OWNED_MEMBER(struct flb_aws_provider *, base_aws_provider);
#endif

    /* prefix */
    OUT_ES_DECLARE_OWNED_MEMBER(flb_sds_t, logstash_prefix);
    OUT_ES_DECLARE_OWNED_MEMBER(flb_sds_t, logstash_prefix_separator);

    /* time key */
    OUT_ES_DECLARE_OWNED_MEMBER(flb_sds_t, time_key);

    /* ... and so on for all other owned members ... */
};
```

And destruction phase is written as:

```c
void flb_elasticsearch_config_destroy(struct flb_elasticsearch_config *config)
{
    if (!config) {
        return;
    }

    /* Free all owned members using the helper macro */
    OUT_ES_DESTROY_OWNED_MEMBER(config, index, flb_free);
    OUT_ES_DESTROY_OWNED_MEMBER(config, type, flb_free);
    OUT_ES_DESTROY_OWNED_MEMBER(config, cloud_user, flb_free);
    OUT_ES_DESTROY_OWNED_MEMBER(config, cloud_passwd, flb_free);

#ifdef FLB_HAVE_AWS
    OUT_ES_DESTROY_OWNED_MEMBER(config, aws_provider, flb_aws_provider_destroy);
    OUT_ES_DESTROY_OWNED_MEMBER(config, base_aws_provider, flb_aws_provider_destroy);
#endif

    OUT_ES_DESTROY_OWNED_MEMBER(config, logstash_prefix, flb_sds_destroy);
    OUT_ES_DESTROY_OWNED_MEMBER(config, logstash_prefix_separator, flb_sds_destroy);
    OUT_ES_DESTROY_OWNED_MEMBER(config, time_key, flb_sds_destroy);

    /* ... destruction for all other owned members ... */

    flb_free(config);
}
```

This approach could be generalized type independent and more readable.

mabrarov committed:
out_es: error handling fix for AWS authentication configuration

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
out_es: memory leak fix when using AWS authentication

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
out_es: NULL pointer dereferencing fix

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov commented on plugins/out_es/es_conf_parse.c:
```diff
@@ -0,0 +1,346 @@
+            if (ec->cloud_passwd == NULL) {
+                flb_utils_split_free(toks);
+                return -1;
+            }
```
Instance of `flb_elasticsearch_config` (which part are `cloud_user` and `cloud_passwd` fields) should be freed anyway - even if it is planned to be used one more time after `flb_es_conf_set_cloud_credentials` function returned non zero exit code - using `flb_es_conf_destroy` function which internally calls `elasticsearch_config_destroy` function which frees memory pointed by `cloud_user` and `cloud_passwd` fields. It is currently implemented in `es_config_ha` and `es_config_simple` functions.

mabrarov committed:
out_es: custom types to track ownership of configuration data

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov commented on plugins/out_es/es_conf_parse.c:
```diff
@@ -0,0 +1,347 @@
+        strcat(cloud_host_buf, port);
+    }
+    return flb_sds_create(cloud_host_buf);
+}
```
It is existing code moved to a new file. Refer to: https://github.com/fluent/fluent-bit/blob/797031cc30d2845b4dcb36cece6d6f2dd363d430/plugins/out_es/es_conf.c#L89
There is no way to fix all issues in this PR without increasing risk of regression due to the lack of comprehensive tests.

mabrarov commented on plugins/out_es/es_conf.c:
```diff
@@ -17,521 +17,885 @@
-    if (!path) {
-        path = "";
+    return 0;
+}
```
It looks like existing issue (existing code moved to another file). Refer to: https://github.com/fluent/fluent-bit/blob/797031cc30d2845b4dcb36cece6d6f2dd363d430/plugins/out_es/es.c#L633

mabrarov commented on plugins/out_es/es_conf.c:
```diff
@@ -17,521 +17,885 @@
-        ctx->index = flb_strdup(f_index->value); /* FIXME */
+        if (ret != 0) {
+            return -1;
+        }
```
Fixed as part of 2816aab4fdaff5940b7d43f7cb1182c330a45017

mabrarov commented on plugins/out_es/es.h:
```diff
@@ -100,51 +103,77 @@ struct flb_elasticsearch {
+struct flb_elasticsearch {
+    /* if HA mode is enabled */
+    int ha_mode;              /* High Availability mode enabled ? */
+    char *ha_upstream;        /* Upstream configuration file      */
```
This member seems to be unused. The same is true for `ha_upstream` member of `flb_forward` struct in plugins/out_forward/forward.h.

mabrarov committed:
out_es: error handling fix when parsing cloud credentials

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov commented on plugins/out_es/es_conf_parse.c:
```diff
@@ -0,0 +1,350 @@
+            if (ec->cloud_passwd.value == NULL) {
+                flb_utils_split_free(toks);
+                return -1;
+            }
```
Returning `-1` prevents silent fall back.

mabrarov commented on plugins/out_es/es_conf.c:
```diff
@@ -17,521 +17,908 @@
+    tmp = flb_upstream_node_get_property(FLB_ES_CONFIG_PROPERTY_BUFFER_SIZE, node);
+    if (tmp) {
+        ec->buffer_size = config_adjust_buffer_size(flb_utils_size_to_bytes(tmp));
+    }
```
This code should be consistent with `flb_config_map_set` function which accepts `-1`:
https://github.com/fluent/fluent-bit/blob/f6c2ae5456436c4426932cc1fc51eac3530a3eaf/src/flb_config_map.c#L725

mabrarov replied:
This new code configures / overrides configuration of upstream node and hence it should act as close to `flb_config_map_set` function as possible. Existing code in master branch (relying on `flb_config_map_set` function) doesn't treat `-1` as error when configuring elasticsearch output plugin. If change of behavior (e.g. making it consistent across all plugins) is required, then it's out of scope of this pull request (which is already too large).

mabrarov commented on plugins/out_es/es_type.h:
```diff
@@ -0,0 +1,413 @@
+ *             If src is NULL, then flb_sds_create is not called,
+ *             @c value of dest is assigned NULL and NULL is returned.
+ *
+ * @return The value assigned to @c value of dest if src is not NULL and
```
Done within b7ce32b41ef5e92536580f538dfe4d42e42b7a57

mabrarov commented on plugins/out_es/es_type.h:
```diff
@@ -0,0 +1,413 @@
+ *             If src is NULL, then flb_strdup is not called, @c value of dest
+ *             is assigned NULL and NULL is returned.
+ *
+ * @return The value assigned to @c value of dest if src is not NULL and
```
Done within b7ce32b41ef5e92536580f538dfe4d42e42b7a57

mabrarov commented:
Hi dear maintainers,

This PR is pending for more than 2.5 years. I appreciate if you could find time and review it again. I believe I have [addressed comment](https://github.com/fluent/fluent-bit/pull/7608#issuecomment-3750223388) of @cosmo0920 which was the only concern at the time the comment was added.

My winter vacation has ended and I won't be able to spend much time on this PR till summer, especially, considering new pending PRs for the same plugin (e.g. #11422 and #11375) - reabasing such changes can be hard.

Thank you.

mabrarov commented:
I believe that similar changes can be implemented / helpful for OpenSearch output plugin (in a dedicated PR).

mabrarov commented on plugins/out_es/es_conf.c:
```diff
@@ -17,521 +17,886 @@
+        return -1;
+    }
+    tmp = flb_upstream_node_get_property(FLB_ES_CONFIG_PROPERTY_LOGSTASH_PREFIX, node);
+    if (tmp && flb_es_sds_copy_str(&ec->logstash_prefix, tmp)) {
```
`flb_upstream_node_get_property` doesn't return `flb_sds_t`. It returns NULL-terminated string, because `flb_sds_t` is transformed into `void *` and then copied into hash table (used by `flb_upstream_node_get_property`) - refer to

https://github.com/fluent/fluent-bit/blob/3e414ac490531b42e5a4a35d19af83152d7b4966/src/flb_hash_table.c#L385

which is called from

https://github.com/fluent/fluent-bit/blob/3e414ac490531b42e5a4a35d19af83152d7b4966/src/flb_hash_table.c#L475

which is called from

https://github.com/fluent/fluent-bit/blob/3e414ac490531b42e5a4a35d19af83152d7b4966/src/flb_upstream_ha.c#L312

which is called from

https://github.com/fluent/fluent-bit/blob/3e414ac490531b42e5a4a35d19af83152d7b4966/src/flb_upstream_ha.c#L433

which is called from `es_config_ha`.

It means that we cannot use `tmp` here without either copying it as `flb_sds_t` (current solution in this PR) or changing the type of `logstash_prefix` field from `flb_sds_t` to another (like `flb_sds_view_t` in https://github.com/fluent/fluent-bit/pull/11647 or `char *`).

mabrarov committed:
out_es: basic support of upstream servers

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
tests: out_elasticsearch: basic support of upstream servers

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
lib: support of tests with callback context depending on config

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
engine_dispatch: support of tests with callback context depending on config

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
out_es: adjustment for new formatter callback

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
tests: out_elasticsearch: adjustment for new formatter callback

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
out_es: support of tests for upstream node configuration properties

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
tests: out_elasticsearch: tests for upstream node configuration properties

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
lib: support of tests with callback context depending on config

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

mabrarov committed:
tests: out_elasticsearch: callback context depending on config

Signed-off-by: Marat Abrarov <abrarov@gmail.com>

edsiper commented:
closing in favor of https://github.com/fluent/fluent-bit/pull/11647

mabrarov commented:
Hi @edsiper,

Is performance tradeoff of #11647 (repeating same operations during every flush) - compared to this PR - considered acceptable? Doesn't it contradict the mission of Fluent Bit?

Thank you.

patrick-stephens commented:
> Hi @edsiper,
> 
> Is performance tradeoff of #11467 (repeating same operations during every flush) - compared to this PR - considered acceptable? Doesn't it contradict the mission of Fluent Bit?
> 
> Thank you.

I think you meant #11647 @mabrarov :)

@cosmo0920 is working on that so may have some comments around it.
